README.md

```diff
@@ -1,11 +1,92 @@
 Rancher Webhook
 ========
- Rancher webhook is both a validating admission webhook and mutating admission webhook for Kubernetes. 
-
+ Rancher webhook is both a validating admission webhook and mutating admission webhook for Kubernetes.
 
 [Explanation of Webhooks in Kubernetes](
 https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
 
+## Background
+The Rancher Webhook is an instance of a Kubernetes admission controller.
+Admission controllers are a standard Kubernetes mechanism to intercept requests to a cluster's
+API server and perform validation or mutation of resources prior to their persistence to
+the cluster database.
+The validation and mutation are performed by custom code whose logic is defined by the instance of an admission controller.
+Cluster administrators may create many such controllers that perform custom validation or mutation in series.
+The Rancher Webhook is one such controller that runs only in the local cluster.
+
+A typical admission controller, such as Rancher's webhook, works as a web-server that receives requests forwarded
+by the main API server. If the Webhook runs its custom validation logic for the resource in question
+and rejects it because the object's structure violates some custom rules, then the whole request to the API server
+is also rejected.
+
+A significant part of the low-level functionality of the Webhook's web-server is implemented
+in [Wrangler](https://github.com/rancher/wrangler). Wrangler provides code for error handling, top-level routing,
+object decoding and response encoding.
+
+The Go web-server itself is configured by [dynamiclistener](https://github.com/rancher/dynamiclistener).
+It handles TLS certificates and the management of associated Secrets for secure communication of other Rancher components with the Webhook.
+
+## Webhook
+The Webhook focuses on the resource matching rules and admission logic for each relevant
+object that needs custom validation. Each object that the Webhook inspects has an `Admit` method which takes a webhook request and response. 
+Its main job is to extract the object from the request,
+perform custom validation or mutation, and modify the response as needed, thus signaling if validation succeeded
+or failed. When the Webhook is done, it returns control to Wrangler which then handles the sending of the verdict
+of the admission controller.
+
+`Admit(webhookResponse *Response, webhookRequest *Request) error`
```

Comment:
```suggestion
`Admit(webhook.Response *Response, webhook.Request *Request) error`
```

Reply:
This will make the names wrong. Also, for simplicity, I did't want to indicate that the types come from Wrangler in this case (or wherever they might get moved soon).